utils/ssh_client.py

```diff
@@ -64,46 +82,38 @@ def connect(self, use_credentials=False, **kwargs):
 
         try:
             if use_credentials:
-                if self.password is None:
-                    message = 'Password is not specified for {}@{}'.format(self.username, self.host)
-                    logger.error(message)
-                    raise SSHConnectionFailed(message)
-                self.client.connect(self.host, username=self.username, password=self.password, **kwargs)
+                self._connect_with_password(**kwargs)
             else:
-                if self.key is None:
-                    message = 'Private key is not specified for {}@{}. Attempt to connect using credentials'.format(
-                        self.username, self.host)
-                    logger.info(message)
-                    self.client.connect(self.host, username=self.username, password=self.password, **kwargs)
-                else:
-                    self.client.connect(self.host, username=self.username, key_filename=self.key,
-                                        passphrase=self.passphrase, **kwargs)
+                try:
+                    self._connect_with_key(**kwargs)
+                except SSHKeyMissing:
+                    self._connect_with_password(**kwargs)
 
         except paramiko.SSHException:
             message = 'Failed to connect to {}@{}'.format(self.username, self.host)
             logger.error(message)
             raise SSHConnectionFailed(message)
 
     def execute(self, command, raise_on_error=False, **kwargs):
-
         if not self.connected:
             self.connect(self.use_credentials)
 
+        logger.info('Executing: {}'.format(command))
         start = time.time()
         stdin, stdout, stderr = self.client.exec_command(command, environment=self.environment, **kwargs)
         r_code = stdout.channel.recv_exit_status()
         exec_time = time.time() - start
+        stdout = ''.join(stdout.readlines())
+        stderr = ''.join(stderr.readlines())
 
+        stdout_log = ''.join(list(line + '\n\t\t' for line in stdout.splitlines())).strip('\n\t\t')
+        stderr_log = ''.join(list(line + '\n\t\t' for line in stderr.splitlines())).strip('\n\t\t')
+        logger.info('Command executed in {} sec\n\tSTDOUT: {}\n\tSTDERR: {}'.format(exec_time, stdout_log, stderr_log))
```

Comment:
Don't change output in logs, log can be single way to investigate and find out some unstable defect. All can be important, event characters between lines.

Reply:
Done